### DISCUSSION TITLE
Is there any way to udpate the header information which is already set?

### DISCUSSION THREAD

**Discussion by samnaction** [curl/curl#7828] [Q&A] [answered]
How to update already set header?

This is already set header. 

```
        struct curl_slist* headers = NULL;

        headers = curl_slist_append(headers, "Accept:");

        headers = curl_slist_append(headers, "Authorization: ABCDC0F725997BEF3C6B90B0E39D9C314DCB58553FA47CD3598BBB8910A8CE6E");

        headers = curl_slist_append(headers, "Host: localhost");

        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
```

After a while, I need to update only the auth token, keeping the other header values intact.

I am doing something like this,
```
// Update the authorization token here
        auto *test = headers;

        while (test != NULL)
        {
            if (strstr(test->data, "Authorization:"))
            {
                strcpy_s(test->data, strlen("Authorization: A5484526D0F7A671ABCD0C091755532A63475719E460ED6A807C4AAD70BCEDA0") + 1, "Authorization: A5484526D0F7A671CE400C091755532A63475719E460ED6A807C4AAD70BCEDA0");
                break;
            }
            test = test->next;
        }
```
The authorization token string will be always same, so strcpy_s should work. But I want to know is there any existing API to update similarly?

**Comment by bagder** [comment]
Sorry, but that's an incomplete code and I don't understand what it tries to do. Can you explain with words how you want it to work?

**Reply by samnaction** [reply]
Updated the question.

**Comment by bagder** [comment, marked as answer]
Make sure to create an updated list of (correct) headers when you are about to start a new transfer!

**Reply by samnaction** [reply]
The Curl_slist_duplicate  is not exposed, so I need to somehow maintain already set  header and update the new header

**Reply by jay** [reply]
> The Curl_slist_duplicate is not exposed, so I need to somehow maintain already set header and update the new header

The correct way is what @bagder said: You'll have to create a new list. The way you are doing it is not recommended. Your code assumes that the old authorization header pointed to by `test->data` has enough space to hold the new authorization header. There's a hacky way around that by making the curl_slist_append for the first auth header a very large string which then afterwards you can go back and fill it with what you want. So it's possible but not recommended.

Also, note curl_slist_append means libcurl is allocating the memory and not your code, therefore if your code uses memory management separate from libcurl (like your application has two separate C runtimes running at the same time, which happens often in Windows) and you were to replace the ->data pointer with your own allocation then later call [curl_slist_free_all](https://curl.se/libcurl/c/curl_slist_free_all.html) to free the list it will attempt to free the memory you allocated and cause a fault because libcurl didn't actually allocate it. There are a number of hacky ways to work around that but again don't.
